src/ILLink.CodeFix/RequiresUnreferencedCodeCodeFixProvider.cs

```diff
@@ -0,0 +1,114 @@
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+
+namespace ILLink.CodeFix
+{
+	[ExportCodeFixProvider (LanguageNames.CSharp, Name = nameof (RequiresUnreferencedCodeCodeFixProvider)), Shared]
+	public class RequiresUnreferencedCodeCodeFixProvider : CodeFixProvider
+	{
+		private const string s_title = "Add RequiresUnreferencedCode attribute to parent method";
+
+		public sealed override ImmutableArray<string> FixableDiagnosticIds
+			=> ImmutableArray.Create (RequiresUnreferencedCodeAnalyzer.DiagnosticId);
+
+		public sealed override FixAllProvider GetFixAllProvider ()
+		{
+			// See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/FixAllProvider.md for more information on Fix All Providers
+			return WellKnownFixAllProviders.BatchFixer;
+		}
+
+		public sealed override async Task RegisterCodeFixesAsync (CodeFixContext context)
+		{
+			var root = await context.Document.GetSyntaxRootAsync (context.CancellationToken).ConfigureAwait (false);
+
+			var diagnostic = context.Diagnostics.First ();
+			var diagnosticSpan = diagnostic.Location.SourceSpan;
+
+			// Find the containing method
+			SyntaxNode targetNode = root!.FindNode (diagnosticSpan);
+			var declaration = targetNode.Parent?.AncestorsAndSelf ()
+				.OfType<MethodDeclarationSyntax> ().FirstOrDefault ();
+
+			if (declaration is not null) {
+				var semanticModel = await context.Document
+					.GetSemanticModelAsync (context.CancellationToken).ConfigureAwait (false);
+				var symbol = semanticModel!.Compilation.GetTypeByMetadataName (
+					RequiresUnreferencedCodeAnalyzer.FullyQualifiedRequiresUnreferencedCodeAttribute);
+
+				// Register a code action that will invoke the fix.
+				context.RegisterCodeFix (
+					CodeAction.Create (
+						title: s_title,
+						createChangedDocument: c => AddRequiresUnreferencedCode (
+							context.Document, root, targetNode, declaration, symbol!, c),
+						equivalenceKey: s_title),
+					diagnostic);
+
+			}
+		}
+
+		private static async Task<Document> AddRequiresUnreferencedCode (
+			Document document,
+			SyntaxNode root,
+			SyntaxNode targetNode,
+			MethodDeclarationSyntax methodDecl,
+			ITypeSymbol requiresUnreferencedCodeSymbol,
+			CancellationToken cancellationToken)
+		{
+			var editor = new SyntaxEditor (root, document.Project.Solution.Workspace);
+			var generator = editor.Generator;
+
+			var semanticModel = await document.GetSemanticModelAsync (cancellationToken).ConfigureAwait (false);
+			if (semanticModel is null) {
+				return document;
+			}
+			var methodSymbol = (IMethodSymbol) semanticModel.GetDeclaredSymbol (methodDecl);
+			var name = semanticModel.GetSymbolInfo (targetNode).Symbol?.Name;
+			SyntaxNode[] attrArgs;
+			if (string.IsNullOrEmpty (name) || HasPublicAccessibility (methodSymbol)) {
+				attrArgs = Array.Empty<SyntaxNode> ();
+			} else {
+				attrArgs = new[] { generator.LiteralExpression ($"calls {name}") };
```

Comment:
I wonder if it should include the type name, particularly if it's a method in another type.

Reply:
I struggled with the particular message to use here.

For adding the type name, shouldn't you be able to see that in the method body? And if you get rid of the codefix you'll see red underlines at the exact source.

I'm not sure if the verbosity helps or hurts here.

Reply:
That's fair - I'd guess in most cases it'll be unambiguous anyway.

Nit: I would capitalize it so the full message for calls to these is:
`Using method 'Method' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Calls "CalleeName".`